### PR TITLE
Shadow Models vs Target Models

### LINKED ISSUE #147: Make get_criteriona nd get_optimizer invisible to user
# Feature
## Desired Behavior / Functionality
Should be invisible

## What Needs to Be Done
Follow Model Handler.

## How Can It Be Tested
Tests should run.

### LINKED ISSUE #158: Consistent loss and acc difference between shadow models and target model
# Issue
## Problem Description
Consistent loss and acc difference between shadow models and target model during training.

## Expected Behavior
For many runs of leakpro, training taget and shadow model from scratch, they should have the same average of loss and acc.
Right now the target model is on average 2-3 % points better than the average of shadow models trained.

## What Needs to be Done
Figure out the difference within the training loop, amount of data, what data the target model is trained on etc.

## How Can It Be Tested or Reproduced
Run leakpro with retrain=True and train and some shadow models for 6 or more epochs a few times

### LINKED ISSUE #270: Custom data transformations
# Issue
## Problem Description
At the moment, we do not support different shadow models to have different data transformations.

## Expected Behavior
Some datasets should be allowed arbitrary transformations

## What Needs to be Done
Fix it.

## How Can It Be Tested or Reproduced
Ensure that the attacks run still.

### LINKED ISSUE #278: Make the user provide eval
# Feature
## Desired Behavior / Functionality
The shadow model handler has implemented its own eval. This is not very flexible and error prone.

## What Needs to Be Done
Make the user provide the eval.

## How Can It Be Tested
Ensure things run.

### LINKED ISSUE #279: Create functionality to create metadata
# Feature
## Desired Behavior / Functionality
Right now it is a bit tricky for the user to create the metadata going into LeakPro

## What Needs to Be Done
Make the creation of metadata invisible to the user

## How Can It Be Tested
Ensure that everything runs

### LINKED ISSUE #281: Users can add fielts to the schemas upon creation
# Issue
## Problem Description
Users should not be able to divert from schema structure

## Expected Behavior
Throw error if user attempts to add fields not within the structure

## What Needs to be Done
Add `model_config = ConfigDict(extra="forbid")  # Prevent extra fields`

## How Can It Be Tested or Reproduced
Ensure user cannot add fields.

### LINKED ISSUE #292: Raise warning/error if user adds parameters in attack config which are not present in attack
# Feature
## Desired Behavior / Functionality
If user supplies for example
`
lira:
   shadow_models: 16
`
Instead of "num_shadow_models" a warning or error should be raised since this field is not part of the attack config.

### PR DESCRIPTION
**Pull request by johanos1**
# Description

Several members have noticed that shadow and target models yield different results. This should not be the case when the shadow model replicates the target model.

There are multiple potential sources:
1. Some examples have two training loops, one for creating the target model and one for providing Leakpro.
2. The optimizer/criterion is not read from metadata of the target model inside the handler.
3. As we utilize subsets to create shadow datasets, the transforms are inherited from the population. Hence, we do not support custom transformations pertaining to the underlying dataset.
4. The data in CIFAR10 seems to be a bit off in the transforms
EDIT: the most impactful error was due to us not creating dataloaders with the same properties in LeakPro as in the training data. More precisely, the shuffle flag. This is now fixed.

Summary of changes

* Updated the AbstractInputHandler to require a UserDataset. This dataset allows LeakPro to create custom transformations in each shadow model dataset and enable the test datasets to use the train transformations.
* Added load functions for criterion and optimizer from the metadata directly in MIAHandler
* Removed the usage of subsets in MIAHandler and instead make use of the UserDataset
* Updated the shadowModelHandler to reflect the UserDataset
* Added eval to the user input and call that in the shadowModelHandler
* Updated the unit tests to use the new metadata configuration and helper functions
* Added safeguards to schemas so users can no longer make changes to a schema

CIFAR10 example
* Updated CIFAR10 data creation and added it directly to the notebook
* Made the UserDataset for CIFAR10

## Resolved Issues

- [ ] fixes #158 
- [ ] fixes #147 
- [ ] fixes #270
- [ ] fixes #279 
- [ ] fixes #278 
- [ ] fixes #281 
- [ ] fixes #292 

## How Has This Been Tested?
Trained 100 shadow models on the same amount of data and compared performance to target, see below. Target i red and shadow models in blue. Shuffle = False in train loader.

Screenshot 2025-03-19 at 13 02 02

This is for shuffle = True using 10 shadow models.
Screenshot 2025-03-19 at 22 36 17

And corresponding LiRA results:
Screenshot 2025-03-19 at 22 37 23

* Unit tests are also running

### CODE REVIEW

**Comment by johanos1** [comment]
We could. The issue was very apparent also for 3 epochs however.

File: leakpro/input_handler/mia_handler.py [hunk 1]
```diff
@@ -228,3 +274,30 @@ def set_test_indices(self:Self, indices:np.ndarray) -> None:
...
+
+    def get_optimizer(self:Self, model:torch.nn.Module) -> None:
+        """Get an instance of same optimizer as used in target model."""
+
+        optimizer_config = self.target_model_metadata.optimizer
+        if not isinstance(optimizer_config, OptimizerConfig):
+            raise ValueError("Optimizer is not a valid schema.")
+
+        # Get all optimizer class names from torch.optim both case sensitive and lower-case
+        optimizer_classes = {
```
Comment (henrikfo) [inline leakpro/input_handler/mia_handler.py:290]: Should also solve the optimizer inheritance issue similar to #242, great!

**Review by henrikfo — APPROVED** [review]
I dont see any apparent issues at all with the code. The "dataset" related changes might temporary impede some examples, but the change and solution is needed. As soon as the issue has been resolved and there is no long a difference between SMs and target models (using a few more epochs) then go ahead and merge!

File: leakpro/leakpro.py [hunk 2]
```diff
@@ -82,6 +85,50 @@ def setup_handler(self:Self, user_input_handler:AbstractInputHandler, configs:di
...
+            dataset_name (str): The name of the dataset
+
+        Returns:
+        -------
+            MIAMetaDataSchema: The metadata for the MIA attack
+
+        """
+
+        return MIAMetaDataSchema(
+            init_params=_get_model_init_params(train_result.model),
```
Comment (fazelehh) [inline leakpro/leakpro.py:119]: Space before and after =

File: leakpro/leakpro.py [hunk 3]
```diff
@@ -82,6 +85,50 @@ def setup_handler(self:Self, user_input_handler:AbstractInputHandler, configs:di
...
         return handler
 
+    @staticmethod
+    def make_mia_metadata(train_result: TrainingOutput,
+                          optimizer: Optimizer,
+                          loss_fn: Module,
+                          dataloader: DataLoader,
+                          test_result: EvalOutput,
+                          epochs: int,
+                          train_indices:list,
```
Comment (fazelehh) [inline leakpro/leakpro.py:95]: lines 95-97 space after :

File: leakpro/tests/input_handler/tabular_utils.py [hunk 4]
```diff
@@ -143,31 +123,27 @@ def create_mock_model_and_metadata(input_size:int) -> str:
...
+    train_result = TrainingOutput(model= model, metrics = EvalOutput(accuracy=0.9, loss=0.1))
+    test_result = EvalOutput(accuracy=0.8, loss=0.2)
+    optimizer = optim.SGD(model.parameters(), lr=parameters.learning_rate)
+    criterion = nn.BCEWithLogitsLoss()
+    train_loader = DataLoader(dataset, batch_size=parameters.batch_size, shuffle=False)
+    train_indices = np.arange(parameters.train_data_points).tolist()
+    test_indices = np.arange(parameters.train_data_points,parameters.train_data_points + parameters.test_data_points).tolist()
+    dataset_name = "Tabular"
+    meta_data = LeakPro.make_mia_metadata(train_result=train_result,
+                                      optimizer=optimizer,
```
Comment (fazelehh) [inline leakpro/tests/input_handler/tabular_utils.py:135]: Space before and after =

**Review by fazelehh — CHANGES_REQUESTED** [review]

File: examples/mia/cifar/cifar_main.ipynb [hunk 5]
```diff
@@ -18,37 +18,119 @@
     "import os\n",
```
Comment (fazelehh) [inline examples/mia/cifar/cifar_main.ipynb:18]: <pre class="ql-syntax" spellcheck="false">Line #17.    test_subset = CifarInputHandler.UserDataset(data[test_indices], targets[test_indices], **train_subset.return_params())
</pre><p>Add a short explanation on why we are using return_params() either as a comment or in the markdown cell </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/269/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (fazelehh) [inline examples/mia/cifar/cifar_main.ipynb:18]: <pre class="ql-syntax" spellcheck="false">Line #24.    train_std = train_subset.data.std(dim=(0, 2, 3))
</pre><p>train_mean and train_std is the same as train_subset.return_params(). Why not using retunr_params() and recalculating mean and std herer?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/269/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (fazelehh) [inline examples/mia/cifar/cifar_main.ipynb:18]: <pre class="ql-syntax" spellcheck="false">Line #33.    train_result.model.to("cpu")
</pre><p>why do you need to bring the model to cpu here and then send the cpu as the device to eval() ? Why not brining the model to cpu right before saving it? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/269/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (fazelehh) [inline examples/mia/cifar/cifar_main.ipynb:18]: <p>I think we can expand on this section. For example, we could mention that attacks can be added or removed from the yaml file. Alos, we could clarify what happens when using LeakPro(CifarInputHandler, ...)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/269/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

File: examples/mia/cifar/cifar_handler.py [hunk 6]
```diff
@@ -61,16 +48,80 @@ def train(
...
+            assert data.max() <= 1.0 and data.min() >= 0.0, "Data should be in range [0,1]"
+
+            self.data = data.float()  # Ensure float type
+            self.targets = targets
+
+            for key, value in kwargs.items():
+                setattr(self, key, value)
+                
+            if not hasattr(self, "mean") or not hasattr(self, "std"):
+                self.mean, self.std = self._compute_mean_std()
```
[the remark this answers is not in the document]
Comment (johanos1) [reply examples/mia/cifar/cifar_handler.py:104]: done. I still need the view though.

File: examples/mia/cifar/cifar_handler.py [hunk 7]
```diff
@@ -45,6 +29,9 @@ def train(
         gpu_or_cpu = device("cuda" if cuda.is_available() else "cpu")
         model.to(gpu_or_cpu)
 
+        accuracy_history = []
```
[the remark this answers is not in the document]
Comment (johanos1) [reply examples/mia/cifar/cifar_handler.py:32]: removed it.

File: examples/mia/cifar/cifar_handler.py [hunk 8]
```diff
@@ -61,16 +48,80 @@ def train(
...
 
                 # Accumulate performance of shadow model
-                train_acc += pred.eq(labels.data.view_as(pred)).sum()
+                train_acc += pred.eq(labels.data.view_as(pred)).sum().item()
                 total_samples += labels.size(0)
                 train_loss += loss.item()
                 
-        avg_train_loss = train_loss / len(dataloader)
-        train_accuracy = train_acc / total_samples  
+            avg_train_loss = train_loss / len(dataloader)
```
[the remark this answers is not in the document]
Comment (johanos1) [reply examples/mia/cifar/cifar_handler.py:55]: done

File: examples/mia/cifar/cifar_handler.py [hunk 9]
```diff
@@ -61,16 +48,80 @@ def train(
...
+        loss, acc = 0, 0
+        with no_grad():
+            for data, target in loader:
+                data, target = data.to(device), target.to(device)
+                target = target.view(-1) 
+                output = model(data)
+                loss += criterion(output, target).item()
+                pred = output.argmax(dim=1) 
+                acc += pred.eq(target).sum().item()
+            loss /= len(loader)
```
[the remark this answers is not in the document]
Comment (johanos1) [reply examples/mia/cifar/cifar_handler.py:78]: done

File: examples/mia/cifar/cifar_handler.py [hunk 10]
```diff
@@ -61,16 +48,80 @@ def train(
                 optimizer.step()
 
                 # Accumulate performance of shadow model
-                train_acc += pred.eq(labels.data.view_as(pred)).sum()
+                train_acc += pred.eq(labels.data.view_as(pred)).sum().item()
```
[the remark this answers is not in the document]
Comment (johanos1) [reply examples/mia/cifar/cifar_handler.py:51]: removed it.

File: leakpro/schemas.py [hunk 11]
```diff
@@ -128,15 +155,13 @@ class MIAMetaDataSchema(BaseModel):
     init_params: Dict[str, Any] = Field(default_factory=dict, description="Model initialization parameters")
     optimizer: OptimizerConfig = Field(..., description="Optimizer configuration")
     loss: LossConfig = Field(..., description="Loss function configuration")
```
[the remark this answers is not in the document]
Comment (johanos1) [reply leakpro/schemas.py:157]: done

File: leakpro/tests/input_handler/tabular_utils.py (hunk 4 above)
Comment (johanos1) [reply leakpro/tests/input_handler/tabular_utils.py:135]: this is test.

File: examples/mia/cifar/cifar_main.ipynb (hunk 5 above)
Comment (johanos1) [reply examples/mia/cifar/cifar_main.ipynb:18]: done

Comment (johanos1) [reply examples/mia/cifar/cifar_main.ipynb:18]: done

Comment (johanos1) [reply examples/mia/cifar/cifar_main.ipynb:18]: added some small text.
